Removing distutils leftover

apraga opened the pull request:
This aims to finish #58073. The patch is untested but follow https://setuptools.pypa.io/en/latest/deprecated/distutils-legacy.html.
I did not touch the install scripts.

Thanks,

Commit:
Removing distutils leftover

keerthanakadiri commented:
Hi @apraga, Can you please sign CLA , thank you!

mihaimaruseac commented:
This looks good on a quick glance, but without signing the CLA we cannot take it.

apraga commented:
With the proper email, it should be good now !

mihaimaruseac reviewed (approved):
Nice! Thank you.

I'll monitor this on the internal imports to make sure it passes all CI. Hopefully we won't have another rollback from a downstream team.

mihaimaruseac commented on third_party/gpus/find_cuda_config.py:
```diff
@@ -65,10 +65,7 @@
-try:
-  from shutil import which
-except ImportError:
-  from distutils.spawn import find_executable as which
+from shutil import which
 # pylint: enable=g-import-not-at-top
```
Linter complains (here and in other places where this pattern occurs): Since now we have a top-level import, it should be sorted in the list of system imports that come above it. Also, the pylint comments are no longer needed.

Can you do these changes please?

apraga replied:
Okay, I've used `shutil.which` instead of `which` to be more consistent with the rest of the imports.

mihaimaruseac commented on tensorflow/api_template_v1.__init__.py:
```diff
@@ -14,8 +14,8 @@
 
-import distutils as _distutils
 import importlib
+import sysconfig
```
For the linter, can you move this to below the `import sys` line?

mihaimaruseac reviewed (changes requested):
There are some changes required by the internal linter. Can you do them, please?

Thank you

Commit:
Correcting import order for linter

Shutil mainly

mihaimaruseac commented:
Oh, now we have a merge conflict to resolve :(

apraga commented:
It should be good to go now !


On Wednesday, August 21st, 2024 at 9:49 PM, Mihai Maruseac ***@***.***> wrote:

> Oh, now we have a merge conflict to resolve :(
> 

> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.

mihaimaruseac commented:
Umm, it seems this has unrelated changes

mihaimaruseac commented on tensorflow/python/kernel_tests/array_ops/manip_ops_test.py:
```diff
@@ -25,12 +25,9 @@
-  # numpy.roll for multiple shifts was introduced in numpy version 1.12.0
-  NP_ROLL_CAN_MULTISHIFT = Version(np.version.version) >= Version("1.12.0")
-except ImportError:
-  NP_ROLL_CAN_MULTISHIFT = False
+from packaging.version import StrictVersion as Version
+# numpy.roll for multiple shifts was introduced in numpy version 1.12.0
```
Can you remove these lines please and move the import immediately after the numpy import?

mihaimaruseac replied:
Sorry, keep the numpy.roll comment, but delete line 31 (import not-at-top)

mihaimaruseac commented:
You still have changes that are unrelated to the PR, you have included commits that already existed on the branch. Unfortunately, this breaks importing the PR into the internal system.

apraga commented:
Ha, I tried to solve the merge conflict above. So what would be the best way to proceed ? Re-ordering the commts ? Undo the merge (but it looked like the CI wasn't run). Thanks

mihaimaruseac commented:
Reordering the commits and removing all the others is the best. But unsure if that would be possible now.

Another alternative would be to close this and recreate a new one

mihaimaruseac commented:
If you open a new PR, please tag me and I'll try to speed it up

Commit:
Fix linter error

apraga commented:
Okay, I've reset the branch unto master et cherry-picked the commit before force-pushing it. Hope it solve the issue :)

mihaimaruseac commented:
This seems to fail with

```
    for scheme in sysconfig.get_scheme_names():
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: module 'tensorflow._api.v2.sysconfig' has no attribute 'get_scheme_names'
```

I think it was passing before the mishap with the bad merge, so maybe conflict resolution removed something that used to work?

mihaimaruseac commented:
This seems to be from a test that does `import tensorflow as tf`. Can you try building a wheel with these changes and see if the import proceeds, please?

apraga commented:
@mihaimaruseac thanks for the follow-up. I tried to build tensorflow from source with different ways (clang18, gcc, clang17 with the binary, clang17 inside the docker image). Each resulted in compilation error. Clang17 on gentoo lead to the farthest but exhaust all the memory 

```[24,203 / 24,842] 16 actions running
    Compiling .../compiler/mlir/tensorflow/transforms/xla_call_module_deserialization.cc; 160s local
    Compiling .../mlir/quantization/stablehlo/passes/quantize_composite_functions.cc; 156s local
    Compiling tensorflow/compiler/mlir/quantization/stablehlo/passes/quantize.cc; 154s local
    Compiling tensorflow/compiler/mlir/tensorflow/transforms/tpu_validate_inputs.cc; 153s local
    Compiling tensorflow/compiler/mlir/quantization/stablehlo/passes/prepare_quantize.cc; 152s local
    Compiling .../compiler/mlir/tensorflow/transforms/drop_while_shape_invariant.cc; 150s local
    Compiling tensorflow/compiler/tf2xla/kernels/fft_ops.cc; 149s local
    Compiling tensorflow/dtensor/mlir/dtensor_send_recv.cc; 148s local ...

Server terminated abruptly (error code: 14, error message: 'Socket closed', log file: '/home/alex/.cache/bazel/_bazel_alex/b43089709cdb88f3f459426ef0ba5527/server/jvm.out')
````

mihaimaruseac commented:
Oh😥 there are two options here: you can try to compile but passing flags to limit the total amount of memory being used by Bazel (`-j 1` might also help), or we can try a hacky test: in an environment where you have TF installed, go to `...dist-packages/tensorflow` and edit all Python files there to not have `distutils` then let's try `import tensorflow as tf` in a Python interpreter started in the same environment, but outside of the `...dist-packages` directory. 

If that passes, I would then check that all edited files are reflected in the PR too

Commit:
Fix conflict between sysconfig

__init__py has :
  from tensorflow._api.v2 import sysconfig
which conflict with python sysconfig

apraga commented:
Okay, I managed to compile tensorflow with `- j 1` and can reproduce the issue. After installation `lib/python3.12/site-packages/tensorflow/__init__.py`, has 2 sysconfig

```python
from tensorflow._api.v2 import sysconfig
import sysconfig
```
Commit below offers a fix (untested, currently compiling).

keerthanakadiri commented:
Hi @apraga, Can you please resolve the conflicts? Thank you!

mihaimaruseac committed:
Merge branch 'master' into no-distutil

apraga commented:
@keerthanakadiri Hi, I've rebasend onto latest master and squashed all commits into one. Hope it's good now !

mihaimaruseac commented:
I took this internally and made a bunch of changes to it (needed some fixed to `BUILD` files and copybara changes). Running the last set of tests now and hopefully it can land by EoD

apraga commented:
Awesome ! Thanks a lot
